Fix numerical instability

iangrooms opened the pull request:
From the beginning we've had problems with numerical stability, i.e. roundoff error, that accumulates during the iterative application of the polynomial filter to data. See, for example, #33, #124, and #135, and the things we've tried to improve this, e.g. #67, #130, and #134.

This PR uses exactly the same polynomial approximation of the target filter, but applies the polynomial to the data in a completely different way. The old way was based on factoring the polynomial in terms of its roots and then iteratively applying each factor to the data. The new method is based on an algorithm for evaluating a polynomial using its coordinates in the Chebyshev basis, the algorithm being based on the three-term recurrence for Chebyshev polynomials.

For now I have left the "Numerical Instability" example notebook in, showing that the new code can filter 1/4 degree vorticity data to 10 degrees with no problems. In fact I have not been able to break the new code in any example I've tried including the Taper filter with very large filter factors.

NoraLoose commented:
Thanks @iangrooms, this sounds really exciting! :rocket:

It is a bit difficult for me to review this PR because there are no notes on how the new method exactly operates. Ian, it would be helpful if you could share notes (if available) and/or provide more comments in the code.  That way, I could try to match up theory and code.

In addition, I am wondering if we can leverage the verification tests that @rabernat set up in https://github.com/ocean-eddy-cpt/gcm-filters/pull/79 (where we would need verifications not only for kernels but for full filter operators). The rationale is the following: This PR swaps out the central filtering algorithm which this package is based on for another algo. Don't we want to test that the new algo leads to the same/similar filtered fields as the old one?

NoraLoose commented:
> I think I'll update the Filter Theory section to describe how it works. 

Sounds great! I will wait with the review until the changes to the filter theory are pushed.

> I guess we could put in some kind of test to compare the result of the new algorithm to the old one, but eventually I think we should just drop the old one and use the new one.

Yes, exactly. As outlined in #79, the idea is to have tests like these:

```
filtered = filter.apply(data)
np.testing.assert_allclose(filtered, expected)
```

where `expected` is a field we save explicitly into the test suite based on the current code.

iangrooms commented:
I added a section on the theory that underpins the new iterative algorithm for applying the filter to data. @NoraLoose can you take a look?

NoraLoose commented:
Looks great @iangrooms! I left some minor comments above.

Two bigger comments are:

1. We should probably delete the numerical instability notebook because it is confusing: it says the filter becomes unstable, but then it just doesn't happen.
2. Since this PR involves major refactoring of the code (the central filter algorithm is essentially swapped out for a new one!), I want to re-iterate what I said above: we need verification tests that the new and old algorithm give the same / similar results. I think we can do this in a similar manner as set up by @rabernat in https://github.com/ocean-eddy-cpt/gcm-filters/pull/79, but we need verifications not only for kernels but for full filter operators. I'm happy to hear other opinions.

iangrooms commented on docs/theory.rst:
```diff
@@ -192,20 +192,52 @@ The minimum number of steps is 3; if ``n_steps`` is not set by the user, or if i
+.. math:: \mathbf{A} = -\left(\frac{2}{s_{\text{max}}}\Delta + \mathbf{I}\right)
+
+be the discrete Laplacian :math:`\Delta` with a rescaling and a shift.
+In principle one could apply the filter to a vector of data :math:`\mathbf{f}` by computing the vectors :math:`T_i(\mathbf{A})\mathbf{f}` and then taking a linear combination with weights :math:`c_i`.
```
Suggestions for how to do that are welcome. We don't want to be so loose with terminology that it's confusing, but at the same time we don't want to be pedantic.

iangrooms commented on gcm_filters/filter.py:
```diff
@@ -201,32 +171,22 @@ def filter_func(field, *args):
-                    field_bar += (
-                        temp_l * 2 * np.real(s_b) / np.abs(s_b) ** 2
-                        + temp_b * 1 / np.abs(s_b) ** 2
+            T_minus_2 = field_bar.copy()
```
I'm responding here for all of the `.copy()` changes. I'm afraid to remove these because Python might just have `T_minus_2` and `field_bar` point to the same thing, such that every time we update `T_minus_2` it also updates `field_bar`, which we don't want.

iangrooms commented:
Thanks for the code review! In response to the two big comments:

1. I agree that we should delete the Numerical Instability notebook before merging the PR. I'm leaving it in for now in case anyone wants to look at it and verify that indeed, the new algorithm does not suffer from numerical instability.
2. I can work to provide some evidence that the old and new algorithms give the same results, at least when the old one is stable. But setting up a new suite of tests to verify that the filters produce the expected results (not necessarily 'the same as the old algorithm', which would be unnecessary if we believe the new algorithm is correct, not to mention being onerous to implement) seems outside the scope of this PR. I think it needs its own issue (cf #79) and its own PR.

NoraLoose commented:
> which would be unnecessary if we believe the new algorithm is correct, not to mention being onerous to implement

The point of the verification tests would be that we don't have to "believe". They would catch any bugs in the derivation / implementation of the new algorithm, which we may have overseen by simply looking at the code.

> I think it needs its own issue (cf https://github.com/ocean-eddy-cpt/gcm-filters/pull/79) and its own PR.

Sure, we can make the verification tests its own issue and PR. But I think those should be resolved first, before we can merge this PR. I can help with these tests.

NoraLoose commented:
Hi @iangrooms, we merged #153 so we can finalize this PR too. Here are some final minor things that have to be done / discussed:

- [ ] Remove numerical instability notebook?
- [ ] What to do about the "Factoring the Gaussian Filter" option (including the docs section)? Do we want to get rid off it because there is no reason to use this functionality, now that numerical instability is not a problem anymore? Or do we want to keep it, to not lose backward compatibility? (If we bump up to version `v0.3`, it is okay to drop the feature as long as we communicate it.) If we keep it, we should add some sentences at the top of the "Factoring the Gaussian Filter" docs section, to the effect of: "Using this feature is not recommended, and only necessary if you run into numerical instability issues, which you most likely won't."
- [ ] Merge the master branch into your branch `three_term_recurrence`, and run the tests **without overwriting the test data**. (You won't overwrite anything if you simply run `pytest` and don't mess with the `GCM_FILTERS_OVERWRITE_TEST_DATA` variable.) If the tests pass, the filter algorithm gives the same result as the previous one.

iangrooms commented:
I'll merge the new tests. In the meantime, I do plan to remove the numerical instability notebook as well as any related comments in the docs, and the 'factoring the Gaussian.' It's a big update, so I think it makes sense to bump the version number. The old version will still be available if anyone needs it.

NoraLoose commented:
If we remove the "Factoring the Gaussian" section in the docs, we should probably also delete the parameter `n_iterations` in the code (+ functionality associated to it). I'm fine with this.

iangrooms commented:
I've removed the numerical instability notebook, removed the factored Gaussian from the docs and code, and merged the latest updates to testing. For some reason building the docs fails (all other tests pass); I noticed that this happened after I pulled in the latest changes from the master branch and before I made my own changes. Not sure how to fix it. Can @NoraLoose or @rabernat take a look?

NoraLoose commented:
I can build the docs locally for this branch. Not sure what is going on either.
